File: android/app/src/main/java/com/pakrinha/khmerphonetic/KeyPadView.kt
package com.pakrinha.khmerphonetic

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.graphics.Typeface
import android.os.Handler
import android.os.Looper
import android.view.HapticFeedbackConstants
import android.view.MotionEvent
import android.view.View

/**
 * The keys, drawn on a canvas and touched directly.
 *
 * The first version made every key a TextView that fired on ACTION_DOWN. That
 * is why the keyboard felt unusable: the letter was committed the instant your
 * finger landed, so a tap a few pixels into the neighbouring key typed the
 * wrong letter, silently, with no way to correct it — and the engine then
 * converted that wrong letter perfectly, which is how "hard to hit" turns into
 * "wrong Khmer comes out".
 *
 * This version behaves like a real keyboard:
 *
 *   - the key commits on *release*, not on touch
 *   - sliding your finger moves the selection, so a bad landing is recoverable
 *   - a preview bubble shows which key is currently selected
 *   - holding a key gets its alternate (the capitals, Khmer digits, a full stop)
 *   - every press gives the standard keyboard haptic
 *
 * Drawing directly also removes the per-keystroke view churn of rebuilding
 * thirty-odd TextViews on every Shift.
 */
@SuppressLint("ViewConstructor")
class KeyPadView(
    context: Context,
    private val khmerFont: Typeface?,
    private val listener: Listener,
) : View(context) {

    interface Listener {
        fun onCharacter(text: String)
        fun onLiteral(text: String)
        fun onAction(action: KeyboardLayout.Act)
    }

    private class Placed(val key: KeyboardLayout.Key, val bounds: RectF)

    private var rows: List<List<KeyboardLayout.Key>> = emptyList()
    private var placed = emptyList<Placed>()
    private var pressed: Placed? = null
    private var longPressFired = false
    private var firedOnDown = false

    private var shifted = false
    private var symbols = false

    private val handler = Handler(Looper.getMainLooper())
    private val preview = KeyPreview(context, khmerFont)

    private val fill = Paint(Paint.ANTI_ALIAS_FLAG)
    private val stroke = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        style = Paint.Style.STROKE
        strokeWidth = 1f
        color = Color.argb(22, 255, 255, 255)
    }
    private val labelPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        textAlign = Paint.Align.CENTER
        color = context.getColor(R.color.kb_text)
    }
    private val hintPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        textAlign = Paint.Align.CENTER
        color = context.getColor(R.color.kb_khmer_hint)
        typeface = khmerFont
    }
    private val altPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        textAlign = Paint.Align.RIGHT
        color = context.getColor(R.color.kb_hint)
        typeface = khmerFont
    }

    private val keyColor = context.getColor(R.color.kb_key)
    private val functionColor = context.getColor(R.color.kb_key_function)
    private val pressedColor = context.getColor(R.color.kb_key_pressed)
    private val accentColor = context.getColor(R.color.kb_accent)

    init {
        isHapticFeedbackEnabled = true
        rows = KeyboardLayout.letterRows(false)
    }

    /* ------------------------------------------------------------ measurement */

    override fun onMeasure(widthSpec: Int, heightSpec: Int) {
        val width = MeasureSpec.getSize(widthSpec)
        setMeasuredDimension(width, (rows.size * ROW_HEIGHT_DP * density).toInt())
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        val gap = KEY_GAP_DP * density
        val rowHeight = ROW_HEIGHT_DP * density
        val usable = (width - paddingLeft - paddingRight).toFloat()
        val list = ArrayList<Placed>()

        rows.forEachIndexed { index, row ->
            val totalWeight = row.sumOf { it.weight.toDouble() }.toFloat()
            var x = paddingLeft.toFloat()
            val top = index * rowHeight
            for (key in row) {
                val w = usable * key.weight / totalWeight
                if (key !is KeyboardLayout.Key.Gap) {
                    list.add(
                        Placed(
                            key,
                            RectF(x + gap, top + gap, x + w - gap, top + rowHeight - gap),
                        )
                    )
                }
                x += w
            }
        }
        placed = list
    }

    /* --------------------------------------------------------------- drawing */

    override fun onDraw(canvas: Canvas) {
        val radius = CORNER_DP * density
        for (p in placed) {
            val key = p.key
            val isFunction = key is KeyboardLayout.Key.Action &&
                key.action != KeyboardLayout.Act.SPACE
            val shiftActive = key is KeyboardLayout.Key.Action &&
                key.action == KeyboardLayout.Act.SHIFT && shifted

            fill.color = when {
                p === pressed -> pressedColor
                shiftActive -> accentColor
                isFunction -> functionColor
                else -> keyColor
            }
            canvas.drawRoundRect(p.bounds, radius, radius, fill)
            canvas.drawRoundRect(p.bounds, radius, radius, stroke)

            val label = labelOf(key)
            val hint = hintOf(key)
            val cx = p.bounds.centerX()

            labelPaint.textSize = (if (isFunction) FUNCTION_TEXT_SP else LABEL_TEXT_SP) * density
            labelPaint.typeface = if (isKhmer(label)) khmerFont else Typeface.DEFAULT

            if (hint == null) {
                canvas.drawText(label, cx, p.bounds.centerY() - (labelPaint.ascent() + labelPaint.descent()) / 2, labelPaint)
            } else {
                // Label sits high, Khmer hint underneath — the same arrangement
                // as the web app's on-screen keyboard.
                canvas.drawText(label, cx, p.bounds.centerY() - 2 * density, labelPaint)
                hintPaint.textSize = HINT_TEXT_SP * density
                hintPaint.typeface = if (isKhmer(hint)) khmerFont else Typeface.DEFAULT
                canvas.drawText(hint, cx, p.bounds.centerY() + HINT_TEXT_SP * density + 2 * density, hintPaint)
            }

            // A small mark in the corner so long-press alternates are findable.
            alternateOf(key)?.let {
                altPaint.textSize = ALT_TEXT_SP * density
                canvas.drawText(it, p.bounds.right - 5 * density, p.bounds.top + ALT_TEXT_SP * density + 3 * density, altPaint)
            }
        }
    }

    private fun labelOf(key: KeyboardLayout.Key): String = when (key) {
        is KeyboardLayout.Key.Char -> key.label
        is KeyboardLayout.Key.Literal -> key.label
        is KeyboardLayout.Key.Action -> key.label
        is KeyboardLayout.Key.Gap -> ""
    }

    private fun hintOf(key: KeyboardLayout.Key): String? = when (key) {
        is KeyboardLayout.Key.Char -> key.hint
        is KeyboardLayout.Key.Literal -> key.hint
        else -> null
    }

    private fun alternateOf(key: KeyboardLayout.Key): String? = when (key) {
        is KeyboardLayout.Key.Char -> key.altChar ?: key.altLiteral
        is KeyboardLayout.Key.Literal -> key.altLiteral
        else -> null
    }

    private fun isKhmer(text: String) = text.any { it.code in 0x1780..0x17FF }

    /* ----------------------------------------------------------------- touch */

    override fun onTouchEvent(event: MotionEvent): Boolean {
        when (event.actionMasked) {
            MotionEvent.ACTION_DOWN -> {
                select(hitTest(event.x, event.y), haptic = true)
                // Backspace is the one key that must act immediately and repeat
                // while held; waiting for release would make deleting painful.
                if (isBackspaceHeld()) {
                    firedOnDown = true
                    listener.onAction(KeyboardLayout.Act.BACKSPACE)
                    scheduleBackspaceRepeat()
                } else {
                    firedOnDown = false
                    scheduleLongPress()
                }
                return true
            }

            MotionEvent.ACTION_MOVE -> {
                val under = hitTest(event.x, event.y)
                if (under !== pressed) {
                    // Slide to correct a bad landing: nothing is typed until you
                    // lift, so the selection can move as much as it likes.
                    select(under, haptic = false)
                    handler.removeCallbacksAndMessages(null)
                    if (under != null) scheduleLongPress()
                }
                return true
            }

            MotionEvent.ACTION_UP -> {
                handler.removeCallbacksAndMessages(null)
                val key = pressed?.key
                if (!longPressFired && !firedOnDown && key != null) fire(key)
                clearPress()
                performClick()
                return true
            }

            MotionEvent.ACTION_CANCEL -> {
                handler.removeCallbacksAndMessages(null)
                clearPress()
                return true
            }
        }
        return super.onTouchEvent(event)
    }

    override fun performClick(): Boolean {
        super.performClick()
        return true
    }

    private fun hitTest(x: Float, y: Float): Placed? {
        placed.firstOrNull { it.bounds.contains(x, y) }?.let { return it }
        // Fingers land in the gaps between keys constantly; snap to the nearest
        // key in the same row rather than swallowing the press.
        return placed.minByOrNull { p ->
            val dx = when {
                x < p.bounds.left -> p.bounds.left - x
                x > p.bounds.right -> x - p.bounds.right
                else -> 0f
            }
            val dy = when {
                y < p.bounds.top -> p.bounds.top - y
                y > p.bounds.bottom -> y - p.bounds.bottom
                else -> 0f
            }
            dx + dy * 4  // vertical misses are much worse than horizontal ones
        }?.takeIf { p ->
            y >= p.bounds.top - SNAP_DP * density && y <= p.bounds.bottom + SNAP_DP * density
        }
    }

    private fun select(target: Placed?, haptic: Boolean) {
        pressed = target
        longPressFired = false
        invalidate()
        if (target == null) {
            preview.hide()
            return
        }
        if (haptic) performHapticFeedback(HapticFeedbackConstants.KEYBOARD_TAP)
        if (showsPreview(target.key)) preview.show(this, target.bounds, labelOf(target.key), hintOf(target.key))
        else preview.hide()
    }

    private fun clearPress() {
        pressed = null
        longPressFired = false
        preview.hide()
        invalidate()
    }

    /** Space, Shift and friends are wide and obvious; a bubble over them is noise. */
    private fun showsPreview(key: KeyboardLayout.Key) =
        key is KeyboardLayout.Key.Char || key is KeyboardLayout.Key.Literal

    private fun scheduleBackspaceRepeat() {
        val repeat = object : Runnable {
            override fun run() {
                if (!isBackspaceHeld()) return
                listener.onAction(KeyboardLayout.Act.BACKSPACE)
                handler.postDelayed(this, REPEAT_INTERVAL_MS)
            }
        }
        handler.postDelayed(repeat, REPEAT_DELAY_MS)
    }

    private fun scheduleLongPress() {
        val target = pressed ?: return
        val alternate = alternateOf(target.key) ?: return
        handler.postDelayed({
            if (pressed === target) {
                longPressFired = true
                performHapticFeedback(HapticFeedbackConstants.LONG_PRESS)
                when (val key = target.key) {
                    is KeyboardLayout.Key.Char ->
                        if (key.altChar != null) listener.onCharacter(key.altChar)
                        else listener.onLiteral(alternate)

                    else -> listener.onLiteral(alternate)
                }
                preview.hide()
            }
        }, LONG_PRESS_MS)
    }

    private fun fire(key: KeyboardLayout.Key) {
        when (key) {
            is KeyboardLayout.Key.Char -> {
                listener.onCharacter(key.label)
                if (shifted) {
                    shifted = false          // one-shot shift, like the web app
                    rows = KeyboardLayout.letterRows(false)
                    requestLayout()
                }
            }

            is KeyboardLayout.Key.Literal -> listener.onLiteral(key.text)

            is KeyboardLayout.Key.Action -> when (key.action) {
                KeyboardLayout.Act.SHIFT -> setLayer(!shifted, symbols)
                KeyboardLayout.Act.LAYER_SYMBOLS -> setLayer(false, true)
                KeyboardLayout.Act.LAYER_LETTERS -> setLayer(false, false)
                else -> listener.onAction(key.action)
            }

            is KeyboardLayout.Key.Gap -> Unit
        }
    }

    private fun setLayer(shift: Boolean, symbol: Boolean) {
        shifted = shift
        symbols = symbol
        rows = if (symbols) KeyboardLayout.symbolRows() else KeyboardLayout.letterRows(shifted)
        requestLayout()
        invalidate()
    }

    /** Back to the plain letter layer, for a fresh input field. */
    fun resetLayers() {
        if (shifted || symbols) setLayer(shift = false, symbol = false)
    }

    /** Backspace repeats while held; the service drives the repeat. */
    fun isBackspaceHeld(): Boolean =
        (pressed?.key as? KeyboardLayout.Key.Action)?.action == KeyboardLayout.Act.BACKSPACE

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        handler.removeCallbacksAndMessages(null)
        preview.hide()
    }

    private val density: Float get() = resources.displayMetrics.density

    companion object {
        // Taller than the first attempt: 52dp rows with 3dp gaps left keys that
        // were hard to hit accurately.
        const val ROW_HEIGHT_DP = 58f
        private const val KEY_GAP_DP = 3.5f
        private const val CORNER_DP = 9f
        private const val LABEL_TEXT_SP = 19f
        private const val FUNCTION_TEXT_SP = 15f
        private const val HINT_TEXT_SP = 12f
        private const val ALT_TEXT_SP = 9f
        private const val SNAP_DP = 14f
        private const val LONG_PRESS_MS = 380L
        private const val REPEAT_DELAY_MS = 420L
        private const val REPEAT_INTERVAL_MS = 55L
    }
}

File: ios/Keyboard/KeyPadView.swift
//
//  The keys, drawn and touched directly.
//
//  Same touch model as the Android app, and for the same reason: firing a key
//  the moment a finger lands makes a keyboard feel broken, because a tap a few
//  points into the neighbouring key types the wrong letter silently — and the
//  engine then converts that wrong letter perfectly.
//
//    - keys commit on release, not on touch
//    - the selection follows the finger, so a bad landing is recoverable
//    - a preview bubble shows which key is selected
//    - holding a key gives its alternate (capitals, Khmer digits, a full stop)
//

import UIKit

protocol KeyPadDelegate: AnyObject {
    func keyPad(_ pad: KeyPadView, didType text: String)
    func keyPad(_ pad: KeyPadView, didEmit literal: String)
    func keyPad(_ pad: KeyPadView, didTrigger action: KeyAction)
}

final class KeyPadView: UIView {

    weak var delegate: KeyPadDelegate?

    /// iPad needs taller rows. A 54pt row is right on a phone but leaves a 13"
    /// iPad with squat, wide keys and a keyboard a fifth the height of the
    /// screen — it looks broken next to the system keyboard.
    static var rowHeight: CGFloat { isPad ? 76 : 54 }

    private static var isPad: Bool { UIDevice.current.userInterfaceIdiom == .pad }

    /// Type scales with the keys, or the labels swim in the middle of them.
    private static var textScale: CGFloat { isPad ? 1.35 : 1 }

    private struct Placed {
        let key: Key
        let frame: CGRect
    }

    private var rows: [[Key]] = KeyboardLayout.letterRows(upper: false)
    private var placed: [Placed] = []
    private var pressedIndex: Int?
    private var longPressFired = false
    private var firedOnDown = false

    private var shifted = false
    private var symbols = false

    private var longPressTimer: Timer?
    private var repeatTimer: Timer?
    private let preview = KeyPreview()

    private let labelFont: UIFont
    private let functionFont: UIFont
    private let hintFont: UIFont
    private let altFont: UIFont

    override init(frame: CGRect) {
        let scale = Self.textScale
        labelFont = .systemFont(ofSize: 19 * scale, weight: .regular)
        functionFont = .systemFont(ofSize: 15 * scale, weight: .regular)
        hintFont = Fonts.khmer(size: 12 * scale)
        altFont = Fonts.khmer(size: 9 * scale)
        super.init(frame: frame)
        backgroundColor = Palette.background
        isMultipleTouchEnabled = false
        contentMode = .redraw
    }

    required init?(coder: NSCoder) { fatalError("not used") }

    var preferredHeight: CGFloat { CGFloat(rows.count) * Self.rowHeight }

    // MARK: - layout

    override func layoutSubviews() {
        super.layoutSubviews()
        let gap: CGFloat = 3.5
        var list: [Placed] = []
        for (index, row) in rows.enumerated() {
            let totalWeight = row.reduce(CGFloat.zero) { $0 + $1.weight }
            var x: CGFloat = 0
            let top = CGFloat(index) * Self.rowHeight
            for key in row {
                let w = bounds.width * key.weight / totalWeight
                if case .gap = key {} else {
                    list.append(
                        Placed(
                            key: key,
                            frame: CGRect(x: x + gap, y: top + gap,
                                          width: w - gap * 2, height: Self.rowHeight - gap * 2)
                        )
                    )
                }
                x += w
            }
        }
        placed = list
        setNeedsDisplay()
    }

    // MARK: - drawing

    override func draw(_ rect: CGRect) {
        guard let context = UIGraphicsGetCurrentContext() else { return }

        for (index, p) in placed.enumerated() {
            let isFunction: Bool
            var isShiftActive = false
            if case let .action(_, action, _) = p.key {
                isFunction = action != .space
                isShiftActive = action == .shift && shifted
            } else {
                isFunction = false
            }

            let fill: UIColor
            if index == pressedIndex { fill = Palette.keyPressed }
            else if isShiftActive { fill = Palette.accent }
            else if isFunction { fill = Palette.keyFunction }
            else { fill = Palette.key }

            let path = UIBezierPath(roundedRect: p.frame, cornerRadius: 9)
            context.setFillColor(fill.cgColor)
            path.fill()
            context.setStrokeColor(UIColor(white: 1, alpha: 0.09).cgColor)
            path.lineWidth = 1
            path.stroke()

            let label = p.key.label
            let hint = p.key.hint
            let font = isFunction ? functionFont : (isKhmer(label) ? Fonts.khmer(size: 19 * Self.textScale) : labelFont)

            if let hint {
                draw(label, in: p.frame, dy: -7 * Self.textScale, font: font, color: Palette.text)
                draw(hint, in: p.frame, dy: 11 * Self.textScale,
                     font: isKhmer(hint) ? hintFont : .systemFont(ofSize: 11 * Self.textScale),
                     color: Palette.khmerHint)
            } else {
                draw(label, in: p.frame, dy: 0, font: font, color: Palette.text)
            }

            if let alternate = p.key.alternate {
                let attributes: [NSAttributedString.Key: Any] = [
                    .font: isKhmer(alternate) ? altFont : UIFont.systemFont(ofSize: 9 * Self.textScale),
                    .foregroundColor: Palette.hint,
                ]
                let size = (alternate as NSString).size(withAttributes: attributes)
                (alternate as NSString).draw(
                    at: CGPoint(x: p.frame.maxX - size.width - 5, y: p.frame.minY + 3),
                    withAttributes: attributes
                )
            }
        }
    }

    private func draw(_ text: String, in frame: CGRect, dy: CGFloat, font: UIFont, color: UIColor) {
        let attributes: [NSAttributedString.Key: Any] = [.font: font, .foregroundColor: color]
        let size = (text as NSString).size(withAttributes: attributes)
        let origin = CGPoint(
            x: frame.midX - size.width / 2,
            y: frame.midY - size.height / 2 + dy
        )
        (text as NSString).draw(at: origin, withAttributes: attributes)
    }

    private func isKhmer(_ text: String) -> Bool {
        text.unicodeScalars.contains { (0x1780...0x17FF).contains(Int($0.value)) }
    }

    // MARK: - touch

    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        guard let point = touches.first?.location(in: self) else { return }
        select(hitTest(point), haptic: true)

        // Backspace is the one key that must act immediately and repeat while
        // held; waiting for release would make deleting painful.
        if isBackspacePressed {
            firedOnDown = true
            delegate?.keyPad(self, didTrigger: .backspace)
            startRepeat()
        } else {
            firedOnDown = false
            scheduleLongPress()
        }
    }

    override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
        guard let point = touches.first?.location(in: self) else { return }
        let under = hitTest(point)
        guard under != pressedIndex else { return }
        // Slide to correct a bad landing: nothing is typed until you lift.
        cancelTimers()
        select(under, haptic: false)
        if under != nil && !isBackspacePressed { scheduleLongPress() }
    }

    override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        cancelTimers()
        if !longPressFired, !firedOnDown, let index = pressedIndex {
            fire(placed[index].key)
        }
        clearPress()
    }

    override func touchesCancelled(_ touches: Set<UITouch>, with event: UIEvent?) {
        cancelTimers()
        clearPress()
    }

    private var isBackspacePressed: Bool {
        guard let index = pressedIndex, case let .action(_, action, _) = placed[index].key else { return false }
        return action == .backspace
    }

    /// Fingers land in the gaps between keys constantly; snap to the nearest key
    /// in the row rather than swallowing the press.
    private func hitTest(_ point: CGPoint) -> Int? {
        if let exact = placed.firstIndex(where: { $0.frame.contains(point) }) { return exact }
        var best: Int?
        var bestDistance = CGFloat.greatestFiniteMagnitude
        for (index, p) in placed.enumerated() {
            let dx = max(p.frame.minX - point.x, 0, point.x - p.frame.maxX)
            let dy = max(p.frame.minY - point.y, 0, point.y - p.frame.maxY)
            guard dy < 14 else { continue }  // vertical misses are much worse
            let distance = dx + dy * 4
            if distance < bestDistance {
                bestDistance = distance
                best = index
            }
        }
        return best
    }

    private func select(_ index: Int?, haptic: Bool) {
        pressedIndex = index
        longPressFired = false
        setNeedsDisplay()
        guard let index else {
            preview.hide()
            return
        }
        if haptic { Click.tap() }
        let p = placed[index]
        switch p.key {
        case .character, .literal:
            preview.show(over: p.frame, in: self, label: p.key.label, hint: p.key.hint)
        default:
            preview.hide()
        }
    }

    private func clearPress() {
        pressedIndex = nil
        longPressFired = false
        firedOnDown = false
        preview.hide()
        setNeedsDisplay()
    }

    private func cancelTimers() {
        longPressTimer?.invalidate()
        longPressTimer = nil
        repeatTimer?.invalidate()
        repeatTimer = nil
    }

    private func startRepeat() {
        repeatTimer = Timer.scheduledTimer(withTimeInterval: 0.42, repeats: false) { [weak self] _ in
            guard let self, self.isBackspacePressed else { return }
            self.repeatTimer = Timer.scheduledTimer(withTimeInterval: 0.055, repeats: true) { [weak self] timer in
                guard let self, self.isBackspacePressed else { timer.invalidate(); return }
                self.delegate?.keyPad(self, didTrigger: .backspace)
            }
        }
    }

    private func scheduleLongPress() {
        guard let index = pressedIndex, placed[index].key.alternate != nil else { return }
        longPressTimer = Timer.scheduledTimer(withTimeInterval: 0.38, repeats: false) { [weak self] _ in
            guard let self, self.pressedIndex == index else { return }
            self.longPressFired = true
            Click.longPress()
            switch self.placed[index].key {
            case let .character(_, _, _, altChar, altLiteral):
                if let altChar { self.delegate?.keyPad(self, didType: altChar) }
                else if let altLiteral { self.delegate?.keyPad(self, didEmit: altLiteral) }
            case let .literal(_, _, _, _, altLiteral):
                if let altLiteral { self.delegate?.keyPad(self, didEmit: altLiteral) }
            default:
                break
            }
            self.preview.hide()
        }
    }

    private func fire(_ key: Key) {
        switch key {
        case let .character(label, _, _, _, _):
            delegate?.keyPad(self, didType: label)
            if shifted {
                shifted = false           // one-shot shift, like the web app
                setLayer(shift: false, symbol: symbols)
            }
        case let .literal(_, text, _, _, _):
            delegate?.keyPad(self, didEmit: text)
        case let .action(_, action, _):
            switch action {
            case .shift: setLayer(shift: !shifted, symbol: symbols)
            case .layerSymbols: setLayer(shift: false, symbol: true)
            case .layerLetters: setLayer(shift: false, symbol: false)
            default: delegate?.keyPad(self, didTrigger: action)
            }
        case .gap:
            break
        }
    }

    private func setLayer(shift: Bool, symbol: Bool) {
        shifted = shift
        symbols = symbol
        rows = symbols ? KeyboardLayout.symbolRows() : KeyboardLayout.letterRows(upper: shifted)
        setNeedsLayout()
        setNeedsDisplay()
    }

    /// Back to the plain letter layer, for a fresh input field.
    func resetLayers() {
        if shifted || symbols { setLayer(shift: false, symbol: false) }
    }

    deinit { cancelTimers() }
}
